common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/RemoteBlockPushResolver.java

```diff
@@ -0,0 +1,959 @@
+          // byte buffer, we cache only the relevant bytes not the entire large buffer to save
+          // memory.
+          ByteBuffer deferredBuf = ByteBuffer.allocate(buf.remaining());
+          deferredBuf.put(buf);
+          deferredBuf.flip();
+          deferredBufs.add(deferredBuf);
+        }
+      }
+    }
+
+    @Override
+    public void onComplete(String streamId) throws IOException {
+      // TODO should the merge manager check for the merge completion ratio here and finalize
+      // TODO shuffle merge if appropriate? So the merge manager can potentially finalize early
+      // TODO and the file channel can be closed even if finalize merge request is somehow not
+      // TODO received from the driver? If so, then we need to know # maps for this shuffle.
+
+      synchronized (partitionInfo) {
+        logger.trace("{} shuffleId {} reduceId {} onComplete invoked",
+          partitionInfo.appShuffleId.appId, partitionInfo.appShuffleId.shuffleId,
+          partitionInfo.reduceId);
+        Map<Integer, AppShufflePartitionInfo> shufflePartitions =
+          mergeManager.partitions.get(partitionInfo.appShuffleId);
+        // When this request initially got to the server, the shuffle merge finalize request
+        // was not received yet. By the time we finish reading this message, the shuffle merge
+        // however is already finalized. We should thus respond RpcFailure to the client.
+        if (shufflePartitions == null || !shufflePartitions.containsKey(partitionInfo.reduceId)) {
+          deferredBufs = null;
+          throw new RuntimeException(String.format("Block %s %s", streamId,
+            ErrorHandler.BlockPushErrorHandler.TOO_LATE_MESSAGE_SUFFIX));
+        }
+        // Check if we can commit this block
+        if (allowedToWrite()) {
+          isWriting = true;
+          // Identify duplicate block generated by speculative tasks. We respond success to
+          // the client in cases of duplicate even though no data is written.
+          if (isDuplicateBlock()) {
+            deferredBufs = null;
+            return;
+          }
+          if (partitionInfo.getCurrentMapIndex() < 0) {
+            writeAnyDeferredBufs();
+          }
+          long updatedPos = partitionInfo.getPosition() + length;
+          boolean indexUpdated = false;
+          if (updatedPos - partitionInfo.getLastChunkOffset() >= mergeManager.minChunkSize) {
+            partitionInfo.updateChunkInfo(updatedPos, mapIndex);
+            indexUpdated = true;
+          }
+          partitionInfo.setPosition(updatedPos);
+          partitionInfo.setCurrentMapIndex(-1);
+
+          // update merged results
+          partitionInfo.blockMerged(mapIndex);
+          if (indexUpdated) {
+            partitionInfo.resetChunkTracker();
+          }
+        } else {
+          deferredBufs = null;
+          throw new RuntimeException(String.format("%s %s to merged shuffle",
```

Comment:
The case for collisions are not tested. 

Had you considered storing the `deferredBufs` of the detected collisions (this side) and write them out right after the collided callback (the other) finished  its writing?

Reply:
There is a `testFailureWith3Streams` which tests collisions

> Had you considered storing the deferredBufs of the detected collisions (this side) and write them out right after the collided callback (the other) finished its writing?
I don't understand the suggestion. The deferredBufs are part of the stream and if the stream doesn't get opportunity to write, then it lands here. We are not storing `deferredBufs` outside the stream.

Reply:
There is a previous conversation related to this [here](https://github.com/apache/spark/pull/30062#discussion_r507403306).

Reply:
I have also added another simple UT for collisions.

Reply:
My idea is sharing those `deferredBufs` in case of collision via the partitionInfo (in-memory and not via a new file, for example stored in a collection to be able to handle multiple collisions too) with the other stream so let this stream finish successfully and the other stream which is able to write to handle this collided `deferredBufs`.

Reply:
Ok. That will impact the memory usage though. Currently these `deferredBufs` live as long as the stream is alive. Storing it outside increases the time they stay in memory and that could increase the memory considerably.

There is another discussion about limiting the number of bufs in memory [here](https://github.com/apache/spark/pull/30062#discussion_r514026014) which is relevant to this topic.
The approach we have used so far with push-merge is that it is best-effort. So, current behavior is in-line with that.
@Victsm @mridulm What do you think?

I do think that if we decide to try this, it should be a follow-up because this change requires us to evaluate how much memory the shuffle service will use.

Reply:
Sure, a follow-up PR is fine for me. My question was really about whether this idea was considered and evaluated. For sure the memory will be increased (at least the time how long this deferred buffs will be kept in the memory but of course as other streams are working this will lead to increased amount too).

Reply:
I have updated this jira 
https://issues.apache.org/jira/browse/SPARK-33331#
I think when we introduce a limit on the number of pending blocks in memory, then we can also add the ignored blocks (the ones that reach here) to memory if there is availability.

Reply:
Have left a comment to the JIRA.

Reply:
Want to add that one of the things we are considering is the necessity of a bounded in-memory buffer on the server side to help with writing the blocks to merged files.
The consideration is more from the aspect of how much potential improvement we could see on the disk write I/O when merging the blocks, and less about reaching better merge ratio for blocks.
These 2 things are however very related though, and will be something that we will consider holistically inside SPARK-33331.